File: csharp/puppeteer-file-download/Scraper.cs
using PuppeteerSharp;
using System.Net.WebSockets;
using System.Text;

class Scraper
{

    private string _auth;

    public Scraper(string auth)
    {
        _auth = auth;
    }

    private async Task<IBrowser> Connect()
    {
        if (_auth == "USER:PASS")
        {
            throw new Exception("Provide Scraping Browsers credentials in AUTH"
                    + " environment variable or update the script.");
        }
        var options = new ConnectOptions()
        {
            BrowserWSEndpoint = "wss://brd.superproxy.io:9222",
            WebSocketFactory = async (uri, options, cToken) =>
            {
                var socket = new ClientWebSocket();
                var authBytes = Encoding.UTF8.GetBytes(_auth);
                var authHeader = "Basic " + Convert.ToBase64String(authBytes);
                socket.Options.SetRequestHeader("Authorization", authHeader);
                socket.Options.KeepAliveInterval = TimeSpan.Zero;
                await socket.ConnectAsync(uri, cToken);
                return socket;
            },
        };
        return await Puppeteer.ConnectAsync(options);
    }

    public async Task Scrape(string url, string selector, string filename)
    {
        Log("Connecting to Browser...");
        var browser = await Connect();
        try
        {
            Log($"Connected! Navigating to {url}...");
            var page = await browser.NewPageAsync();
            var client = await page.Target.CreateCDPSessionAsync();
            await page.GoToAsync(url, /* timeout= */ 2 * 60 * 1000);
            Log("Navigated! Initiating download...");
            var requestId = await InitiateDownload(page, client, selector);
            Log($"Download started! Stream it to {filename}...");
            var stream = await client.SendAsync("Fetch.takeResponseBodyAsStream", new
            {
                requestId = requestId,
            });
            var file = File.OpenWrite(filename);
            var fileSize = 0;
            while (true)
            {
                var chunk = await client.SendAsync("IO.read", new
                {
                    handle = stream["stream"],
                });
                var data = (bool) chunk["base64Encoded"]!
                    ? Convert.FromBase64String((string) chunk["data"]!)
                    : Encoding.UTF8.GetBytes((string) chunk["data"]!);
                await file.WriteAsync(data);
                fileSize += data.Length;
                if ((bool) chunk["eof"]!)
                {
                    break;
                }
            }
            file.Close();
            Log($"Download saved! Size {fileSize}.");
        }
        finally
        {
            await browser.CloseAsync();
        }
    }

    private async Task<string> InitiateDownload(IPage page, ICDPSession client, string selector)
    {
        var task = new TaskCompletionSource<string>();
        void onRequestPaused(object? sender, MessageEventArgs args)
        {
            if (args.MessageID == "Fetch.requestPaused")
                task.TrySetResult((string) args.MessageData["requestId"]!);
        }
        void onClick(Task click)
        {
            if (click.IsFaulted)
                task.TrySetException(click.Exception!);
        }
        await client.SendAsync("Fetch.enable", new
        {
            patterns = new[]{
                new { requestStage = "Response", resourceType = "Document" },
            },
        });
        client.MessageReceived += onRequestPaused;
        var _ = page.ClickAsync(selector).ContinueWith(onClick);
        return await task.Task;
    }

    private static string Env(string name, string defaultValue)
    {
        return Environment.GetEnvironmentVariable(name) ?? defaultValue;
    }

    private static void Log(string message)
    {
        Console.WriteLine(message);
    }

    public static async Task Main()
    {
        var auth = Env("AUTH", "USER:PASS");
        var url = Env("TARGET_URL", "https://myjob.page/tools/test-files");
        var selector = Env("SELECTOR", "a[role=button]");
        var filename = Env("FILENAME", "./testfile.zip");
        var scraper = new Scraper(auth);
        await scraper.Scrape(url, selector, filename);
    }

}

File: java/playwright-file-download/src/main/java/Scraper.java
import com.microsoft.playwright.*;
import com.google.gson.*;
import java.util.HashMap;
import java.util.Base64;
import java.io.File;
import java.io.FileOutputStream;
import java.util.concurrent.CompletableFuture;

public class Scraper {

    private Playwright _pw;
    private String _auth;

    public Scraper(Playwright pw, String auth) {
        _pw = pw;
        _auth = auth;
    }

    private Browser connect() throws Exception {
        if (_auth == "USER:PASS") {
            throw new Exception("Provide Scraping Browsers credentials in AUTH"
                    + " environment variable or update the script.");
        }
        var endpointURL = "wss://" + _auth + "@brd.superproxy.io:9222";
        return _pw.chromium().connectOverCDP(endpointURL);
    }

    public void scrape(String url, String selector, String filename) throws Exception {
        log("Connecting to Browser...");
        var browser = connect();
        try {
            log("Connected! Navigating to %s...", url);
            var page = browser.newPage();
            var client = page.context().newCDPSession(page);
            page.navigate(url);
            log("Navigated! Initiating download...");
            var requestId = initiateDownload(page, client, selector);
            log("Download started! Stream it to %s...", filename);
            var stream = client.send("Fetch.takeResponseBodyAsStream", fetchStreamParams(requestId))
                .get("stream")
                .getAsString();
            var file = new File(filename);
            var fileStream = new FileOutputStream(file);
            var fileSize = 0;
            while (true) {
                var chunk = client.send("IO.read", ioReadParams(stream));
                var data = chunk.get("base64Encoded").getAsBoolean()
                    ? Base64.getDecoder().decode(chunk.get("data").getAsString())
                    : chunk.get("data").getAsString().getBytes("UTF-8");
                fileSize += data.length;
                fileStream.write(data);
                if (chunk.get("eof").getAsBoolean()) {
                    break;
                }
            }
            fileStream.close();
            log("Download saved! Size %d.", fileSize);
        } finally {
            browser.close();
        }
    }

    private String initiateDownload(Page page, CDPSession client, String selector) throws Exception {
        var task = new CompletableFuture<String>();
        client.send("Fetch.enable", fetchEnableParams());
        client.on("Fetch.requestPaused", event -> {
            task.complete(event.get("requestId").getAsString());
        });
        CompletableFuture.runAsync(() -> {
            try {
                page.click(selector);
            } catch (Exception e) {
                task.completeExceptionally(e);
            }
        });
        return task.get();
    }

    private static JsonObject fetchStreamParams(String requestId) {
        var params = new JsonObject();
        params.addProperty("requestId", requestId);
        return params;
    }

    private static JsonObject ioReadParams(String handle) {
        var params = new JsonObject();
        params.addProperty("handle", handle);
        return params;
    }

    private static JsonObject fetchEnableParams() {
        var params = new JsonObject();
        var patterns = new JsonArray();
        var pattern = new JsonObject();
        pattern.addProperty("requestStage", "Response");
        pattern.addProperty("resourceType", "Document");
        patterns.add(pattern);
        params.add("patterns", patterns);
        return params;
    }

    private static void log(String message, Object... params) {
        System.out.println(String.format(message, params));
    }

    public static void main(String[] args) throws Exception {
        var env = System.getenv();
        var auth = env.getOrDefault("AUTH", "USER:PASS");
        var targetUrl = env.getOrDefault("TARGET_URL", "https://calmcode.io/datasets/bigmac");
        var selector = env.getOrDefault("SELECTOR", "button.border");
        var filename = env.getOrDefault("FILENAME", "./testfile.csv");
        var options = new Playwright.CreateOptions()
            .setEnv(new HashMap<>(env){{
                put("PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD", "1");
            }});
        try (var pw = Playwright.create(options)) {
            var scraper = new Scraper(pw, auth);
            scraper.scrape(targetUrl, selector, filename);
        }
    }

}
